lisa-sets/src/main/scala/lisa/maths/settheory/functions/Functionals.scala

```diff
@@ -0,0 +1,1365 @@
+        lastStep,
+        existentialEquivalenceDistribution of (P -> lambda(a, in(pair(a, t), s)), Q -> lambda(a, ((a === x) /\ (t === y))))
+      )
+
+      val exRedundant = have(exists(a, ((a === x) /\ (t === y))) <=> (t === y)) subproof {
+        have((a === x) /\ (t === y) |- (t === y)) by Restate
+        val fwd = thenHave(exists(a, (a === x) /\ (t === y)) |- (t === y)) by LeftExists
+
+        have((t === y) |- (x === x) /\ (t === y)) by Restate
+        val bwd = thenHave((t === y) |- exists(a, (a === x) /\ (t === y))) by RightExists
+
+        have(thesis) by Tautology.from(fwd, bwd)
+      }
+
+      have((t === y) <=> in(t, singleton(y))) by Restate.from(singletonHasNoExtraElements of (y -> t, x -> y))
+      have(in(t, functionRange(s)) <=> in(t, singleton(y))) by Tautology.from(lastStep, exRedundant, exToEq, inDom)
+      thenHave(forall(t, in(t, functionRange(s)) <=> in(t, singleton(y)))) by RightForall
+      have(thesis) by Tautology.from(lastStep, extensionalityAxiom of (x -> functionRange(s), y -> singleton(y)))
+    }
+
+    have(functional(s)) by Tautology.from(relS, uniq, functional.definition of f -> s)
+
+    have(thesis) by Tautology.from(ran, dom, lastStep)
+
+  }
+
+  val setOfFunctionsUniqueness = Theorem(
+    ∃!(z, ∀(t, in(t, z) <=> (in(t, powerSet(cartesianProduct(x, y))) /\ functionalOver(t, x))))
+  ) {
+    have(thesis) by UniqueComprehension(powerSet(cartesianProduct(x, y)), lambda(t, functionalOver(t, x)))
+  }
+
+  /**
+   * Set of functions --- All functions from `x` to `y`, denoted `x → y` or
+   * `→(x, y)`.
+   *
+   * Since functions from `x` to `y` contain pairs of the form `(a, b) | a ∈
+   * x, b ∈ y`, it is a filtering on the power set of their product, i.e. `x
+   * → y ⊆ PP(x * y)`.
+   */
+  val setOfFunctions = DEF(x, y) --> The(z, ∀(t, in(t, z) <=> (in(t, powerSet(cartesianProduct(x, y))) /\ functionalOver(t, x))))(setOfFunctionsUniqueness)
+  val |=> = setOfFunctions
+
+  extension (x: Term) {
+    // Infix notation for a set of functions: x |=> y
+    def |=>(y: Term): Term = setOfFunctions(x, y)
+  }
+
+  // class SetOfFunctions(val x: Term, val y: Term) extends AppliedFunctional(setOfFunctions, Seq(x, y)) with LisaObject[SetOfFunctions] {
+  //   override def substituteUnsafe(map: Map[lisa.fol.FOL.SchematicLabel[?], lisa.fol.FOL.LisaObject[?]]): SetOfFunctions = SetOfFunctions(x.substituteUnsafe(map), y.substituteUnsafe(map))
+
+  //   override def toString(): String = x.toStringSeparated() + " |=> " + y.toStringSeparated()
+  //   override def toStringSeparated(): String = toString()
+  // }
+  // object SetOfFunctions {
+  //   def unapply(sof: SetOfFunctions): Option[(Term, Term)] = sof match
+  //     case AppliedFunctional(label, Seq(x, y)) if label == setOfFunctions => Some((x, y))
+  //     case _ => None
+  // }
+  // extension (x: Term) {
+  //   // Infix notation for a set of functions: x |=> y
+  //   def |=>(y: Term): Term = SetOfFunctions(x, y)
+  // }
```

Comment:
There is one regression: in the type package `|=>` was used as the setOfFunctions, but it was declared as an infix which gave nicer printing. I tried to replicate this here but I failed. The `with LisaObject[SetOfFunctions]` part was raising an error of `Type argument lisa.maths.settheory.functions.Functionals.SetOfFunctions does not conform to upper bound Common.this.LisaObject[lisa.maths.settheory.functions.Functionals.SetOfFunctions]`